### ISSUE TITLE
Duplicate entries when creating child documents at the same time as parent documents

### ISSUE THREAD

**Issue by chrisfosterelli** [neumino/thinky#166]
Hi there!

I noticed that the following appears to work as desired:

``` 
var course = new Course({
  ...
  sections : []
});
course.saveAll().then(function() {
  course.sections.push(new Section({ ... }));
  course.sections.push(new Section({ ... }));
  course.saveAll().then(function(result) {
    res.send(result);
  });
});
```

But if I do the below, it does not:

```
var course = new Course({
  ...
  sections : [
    new Section({ ... }),
    new Section({ ... })
  ]
});
course.saveAll().then(function() {
  res.send(result);
});
```

The latter example results in four `Section` documents being created. Two appear to be missing the relational field (`courseId`), while the other two are the actual documents I wanted.

I may be misunderstanding something, so any help is appreciated!

**Comment by neumino** [comment]
What version of thinky are you using?
Can you also provide the relations between `Course` and `Section`? I believe it's a `hasMany`/`belongsTo` since you have a `courseId` field. There's a test for this case here:
https://github.com/neumino/thinky/blob/master/test/advanced.js#L208

Can you provide a snippet (with the model/relations) to reproduce this bug?

Thanks!

**Comment by chrisfosterelli** [comment]
Sorry for the wait @neumino! My version of thinky is 1.15.7.

You are correct about the relation being a `hasMany`/`belongsTo`. 
I've attached a more detailed reproduction script below for you:

```javascript
var thinky = require('thinky')();

var Parent = thinky.createModel('Parent', {
  id    : String,
  attr1 : String
});

var Child = thinky.createModel('Child', {
  id    : String,
  attr2 : String
});

Parent.hasMany(Child, "children", "id", "parentId");
Child.belongsTo(Parent, "parent", "parentId", "id");

var par = new Parent({
  attr1 : 'Test',
  children : [
    new Child({ attr2 : 'Test2' }),
    new Child({ attr2 : 'Test3' })
  ]
});

par.saveAll().then(function(result) {
  console.log(result);
});
```

The output is the following:

```
{ attr1: 'Test',
  children: 
   [ { attr2: 'Test2',
       id: '2fd392d2-5c89-4edb-be3c-f85e584e8195',
       parentId: '10d9c542-7dfb-4ca6-8909-b449a6d94dd4' },
     { attr2: 'Test3', id: '11ba7c32-02ca-4843-8c8b-359a4bd923ac' } ],
  id: '10d9c542-7dfb-4ca6-8909-b449a6d94dd4' }
```

But in the database, you can see the following:

![image](https://cloud.githubusercontent.com/assets/2072672/5699428/5d8765ea-99dc-11e4-91a9-c9940b9c7b73.png)

Hope that helps, let me know if you need any additional information!

**Comment by neumino** [comment]
Ah thanks a lot.
I'm not quite sure what's going on yet, but at least I know why the tests were passing.

If you declare `par` this way, things work as expected:
```
var par = new Parent({
  attr1 : 'Test',
});
par.children = [
  new Child({ attr2 : 'Test2' }),
  new Child({ attr2 : 'Test3' })
]
```

I'll keep looking at what's going on here.

**Comment by chrisfosterelli** [comment]
Ah, okay great! I will use that as a better workaround than mine in the mean time, as it doesn't require saving twice.

**Comment by neumino** [comment]
Ok, I actually found the bug. we shouldn't save stuff in `__proto__._hasMany` until they are actually saved. There are a few possible solutions, I'll think a bit more about which one to implement.
- We can delay setting `_hasMany` by putting it in `setSaved`
- Move the code that save things in `_hasMany` in its own function and call it in `model`

**Comment by neumino** [comment]
Hello @chrisfosterelli!

Thanks for reporting this nasty bug and for your patience. The fix was released in 1.15.8

**Comment by chrisfosterelli** [comment]
Awesome, thanks for the fix @neumino!
